asteroid/dsp/beamforming.py

```diff
@@ -248,6 +248,91 @@ def compute_beamforming_vector(target_scm: torch.Tensor, noise_scm: torch.Tensor
         return bf_vect
 
 
+class GEVDBeamformer(Beamformer):
+    """Generalized eigenvalue decomposition speech distortion weighted multichannel Wiener filter.
+
+        Compare to SDW-MWF, spatial covariance matrix are computed from low rank approximation
+        based on eigen values decomposition,
+        see equation 62 in `[1] <https://hal.inria.fr/hal-01390918/file/14-1.pdf>`_.
+
+    Attributes:
+        mu (float): Speech distortion constant.
+        rank (int): Rank for the approximation of target covariance matrix,
+            no approximation is made if rank is None.
+
+    References:
+        [1] R. Serizel, M. Moonen, B. Van Dijk and J. Wouters,
+        "Low-rank Approximation Based Multichannel Wiener Filter Algorithms for
+        Noise Reduction with Application in Cochlear Implants,"
+        in IEEE/ACM Transactions on Audio, Speech, and Language Processing, April 2014.
+    """
+
+    def __init__(self, mu: float, rank: int):
+        self.mu = mu
+        self.rank = rank
+
+    def compute_beamforming_vector(self, target_scm: torch.Tensor, noise_scm: torch.Tensor):
+        """Compute beamforming vectors for GEVD beamFormer.
+
+        Args:
+            target_scm (torch.ComplexTensor): shape (batch, mics, mics, freqs)
+            noise_scm (torch.ComplexTensor): shape (batch, mics, mics, freqs)
+
+        Returns:
+            torch.ComplexTensor: shape (batch, mics, freqs)
+
+        """
+        #  GEV decomposition of noise_scm^(-1) * target_scm
+        e_values, e_vectors = _generalized_eigenvalue_decomposition(
+            target_scm.permute(0, 3, 1, 2),  # bmmf -> bfmm
+            noise_scm.permute(0, 3, 1, 2),  # bmmf -> bfmm
+        )
+
+        #  Prevent negative and infinite eigenvalues
+        eps = torch.finfo(e_values.dtype).eps
+
+        e_values = torch.maximum(e_values, eps * torch.ones(e_values.shape))
+        e_values = torch.minimum(e_values, 1e6 * torch.ones(e_values.shape))
+
+        #  Sort eigen values and vectors in descending-order
+        e_values = torch.diag_embed(torch.flip(e_values, [-1]))
+        e_vectors = torch.flip(e_vectors, [-1])
+
+        #  Force zero values for all GEV but the highest
+        if self.rank:
+            e_values[..., self.rank :, :] = 0.0
+
+        #  Compute bf vectors as SDW MWF filter  in eigen space
+        complex_type = e_vectors.dtype
+        inv_ev_plus_mu = torch.linalg.inv(
+            e_values + self.mu * torch.eye(e_values.shape[-1]).expand_as(e_values)
+        ).to(complex_type)
+        bf_vect = (
+            e_vectors @ e_values.to(complex_type) @ inv_ev_plus_mu @ torch.linalg.inv(e_vectors)
+        )
```

Comment:
I'm unsure, but could this be computed with a `solve` call rather than `inv`? It's usually much faster, and lighter in memory.

Reply:
`solve` is advised to replace a multiplication on the left by the inverse: `torch.linalg.solve(A, B) == A.inv() @ B`
[see pytorch doc](https://pytorch.org/docs/stable/generated/torch.linalg.inv.html#torch.linalg.inv).
But in that case, I didn't find a way to replace `inv` by `solve`. But I suggest to use `inv` only once, this seems quicker.

Reply:
Perfect